src/openloco/windows/terraform.cpp

```diff
@@ -19,13 +77,2267 @@ namespace openloco::ui::windows::terraform
+
+                    if (!count)
+                        return;
+                }
+            }
+            else
+            {
+                if (!(_mapSelectionFlags & 1))
+                    return;
+            }
+
+            uint32_t raiseCost = 0;
+            uint32_t lowerCost = 0;
+
+            if (is_editor_mode() || xPos == 0x8000)
+            {
+                raiseCost = 0x80000000;
+                lowerCost = 0x80000000;
+            }
+            else
+            {
+                lowerCost = lowerLand(GameCommandFlag::flag_2);
+
+                raiseCost = raiseLand(GameCommandFlag::flag_2);
+            }
+            setAdjustCost(raiseCost, lowerCost);
+        }
+
+        // 0x004BC9ED
+        static void on_tool_down(window& self, const widget_index widgetIndex, const int16_t x, const int16_t y)
+        {
+            if (widgetIndex != common::widx::panel)
+                return;
+            if (!(_mapSelectionFlags & 1))
+                return;
+            if (is_editor_mode())
+            {
+                if (_adjustToolSize != 0)
+                {
+                    if (_lastSelectedLand != 0xFF)
+                    {
+                        _gGameCommandErrorTitle = string_ids::error_cant_change_land_type;
+                        game_commands::do_24(_mapSelectionAX, _mapSelectionAY, _mapSelectionBX, _mapSelectionBY, _lastSelectedLand, GameCommandFlag::apply);
+                    }
+                }
+            }
+            else
+            {
+                // CHANGE: Allows the player to change land type outside of the scenario editor.
+                if (_adjustToolSize != 0)
+                {
+                    if (_lastSelectedLand != 0xFF)
+                    {
+                        _gGameCommandErrorTitle = string_ids::error_cant_change_land_type;
+                        game_commands::do_24(_mapSelectionAX, _mapSelectionAY, _mapSelectionBX, _mapSelectionBY, _lastSelectedLand, GameCommandFlag::apply);
+                    }
+                }
+            }
+
+            _currentTool = 3;
+        }
+
+        // 0x004BC9E2
+        static void event_12(window& self, const widget_index widgetIndex, const int16_t x, const int16_t y)
+        {
+            if (widgetIndex != common::widx::panel)
+                return;
+
+            auto window = WindowManager::findAt(x, y);
+            if (window == nullptr)
+                return;
+
+            widget_index newWidgetIndex = window->find_widget_at(x, y);
+            if (newWidgetIndex == -1)
+                return;
+
+            auto widget = window->widgets[newWidgetIndex];
+            if (widget.type != widget_type::viewport)
+                return;
+
+            auto viewport = window->viewports[0];
+            if (viewport == nullptr)
+                return;
+
+            auto zoom = viewport->zoom;
+
+            int16_t dx = static_cast<int16_t>(0xFFF0);
+            dx /= pow(2, zoom);
```

Comment:
Use `std::pow` with <cmath> header.

Reply:
Std::pow is a floating point calc. We should avoid them as much as possible and use dx /= (1 << zoom)

Reply:
I was thinking shifting too, but that caused UB for signed variables.

Reply:
Do we really need to over-optimise stuff in windows?

Reply:
Perhaps we can shift a positive number and then negate it.

Reply:
Why not just `dx >>= zoom`?

Reply:
@IntelOrca Because it's signed. Undefined behaviour.

Reply:
I mean it's not important for the ui but may as well just make it a blanket enforcement until we split off the ui into a separate project.

Reply:
@AaronVanGeffen only if the number is negative. I don't understand why C++ and many other languages don't have a separately defined arithmetic and logical shift operator.  Not to mention not having rotate shift operators either. Most architectures have the instruction, and the compiler can easily implement one if the architecture doesn't support it.